browser/views.py

```diff
@@ -624,3 +625,63 @@ def attendee_create(request, course_id):
         'form': form,
     }
     return render(request, template, context)
+
+
+@staff_member_required
+def add_investigator_record(request, person_id):
+    person = get_object_or_404(Person, pk=person_id)
+    template = "browser/investigator.html"
+    form = InvestigatorForm()
+
+    if request.method == 'POST':
+        form = InvestigatorForm(request.POST, instance=person)
+        if form.is_valid():
+            investigator = Investigator(subject=form.cleaned_data.get('subject'),
+                                            role=form.cleaned_data.get('role'),
+                                            person_id=person_id, year=form.cleaned_data.get('year'),
+                                            changed_by=request.user)
+            investigator.save()
+            return HttpResponseRedirect(reverse('person', args=(person.id,)))
+
+    context = {
+        'form': form,
+        'person': person,
+        'type': 'create_investigator'
+    }
+
+    return render(request, template, context)
+
+
+@staff_member_required
+def edit_investigator_record(request,person_id,research_id):
+    person = get_object_or_404(Person, pk=person_id)
+    research = get_object_or_404(Investigator, pk=research_id)
+    template = "browser/investigator.html"
+
+    # Retreive investigator data already stored in database and display to the user
+    form = InvestigatorForm(initial={'subject': research.subject, 'role': research.role,
+                                             'year': research.year})
+
+    if request.method == 'POST':
+        form = InvestigatorForm(request.POST, instance=person)
+        if form.is_valid():
+            research.subject = form.cleaned_data.get('subject')
+            research.role = form.cleaned_data.get('role')
+            research.year = form.cleaned_data.get('year')
+            research.save()
+            return HttpResponseRedirect(reverse('person', args=(person.id,)))
+
+    context = {
+        'form': form,
+        'person': person,
+        'investigator_data': research,
+        'type': 'edit_investigator'
+    }
+
+    return render(request, template, context)
+
+
+@staff_member_required
+def delete_investigator_record(request, person_id):
+    Investigator.objects.filter(id=request.POST.get("investigator_delete_modal_id"), person_id=person_id).delete()
```

Comment:
do you need the person id for this? isn't the investigator id a primary key?

Reply:
also, what if someone makes a GET request to this method?

Reply:
Yes, person_id is not required for deletion because investigator id is the primary key. I have edited the delete method wherein only the research id or investigator id is required.

I logged out of the system and made this request <b>"GET /person/2303/investigator/37804/delete/ HTTP/1.1"</b>. And, I was not able to delete the investigator record and was redirected to the admin page.

Reply:
the problem here is not being logged in or logged out. it's that you assume only POST requests are made to this method. If I'm logged in and make a GET request, then your code will fail most likely.

Reply:
I have added <b>if request.method is post </b> check in the delete method